hw/drivers/uart/max3107/src/max3107.c

```diff
@@ -0,0 +1,1059 @@
+    for (factor_ix = 0; factor_ix < max_factor; ++factor_ix) {
+        /*
+         * Pre divider does not apply when PLL is not used. Set lower and upper
+         * limits to 1 to have same code for PLL/non PLL.
+         */
+        if (factor_ix == 0) {
+            pre_div_min = 1;
+            pre_div_max = 1;
+        } else {
+            /* Lower and upper frequency limits used to get pre divider range */
+            pre_div_max = clockf / fpllin_min[factor_ix];
+            pre_div_min = (clockf + fpllin_max[factor_ix] - 1) / fpllin_max[factor_ix];
+            /* Make sure divider is in correct range. */
+            pre_div_min = min(63, pre_div_min);
+            pre_div_min = max(1, pre_div_min);
+            pre_div_max = min(63, pre_div_max);
+            pre_div_max = max(1, pre_div_max);
+        }
+        /* Loop for x1, x2 and x4 modes. */
+        for (mul = 1; mul <= mode_mul; mul <<= 1) {
+            for (pre_div = pre_div_min; pre_div <= pre_div_max; ++pre_div) {
+                fref = (clockf / pre_div) * factors[factor_ix];
+                div = (fref * mul + (br / 2)) / br;
+                div = max(div, 16);
+
+                actual_br = mul * fref / div;
+                if (abs((int)(actual_br - br)) < abs((int)(best_br - br))) {
+                    best_br = actual_br;
+                    cfg->div_lsb = (uint8_t)(div >> 4);
+                    cfg->div_msb = (uint8_t)(div >> 12);
+                    cfg->brg_config = (div & 0xF) | ((mul == 4) ? 0x20 : ((mul == 2) ? 0x10 : 0));
+                    /* If best choice is factor_ix == 0, no need for PLL */
+                    if (factor_ix == 0) {
+                        cfg->clk_source |= CLCSOURCE_PLLBYPASS;
+                        cfg->clk_source &= ~CLCSOURCE_PLLEN;
+                    } else {
+                        cfg->pll_config = pre_div | ((factor_ix - 1) << 6);
+                        cfg->clk_source &= ~CLCSOURCE_PLLBYPASS;
+                        cfg->clk_source |= CLCSOURCE_PLLEN;
+                    }
+                }
+            }
+        }
+    }
+
+    return best_br;
+}
+
+int
+max3107_config_uart(struct max3107_dev *dev, const struct uart_conf_port *conf)
+{
+    int rc;
+
+    if (dev->cfg.crystal_en) {
+        dev->regs.clock.clk_source |= CLCSOURCE_CRYSTALEN;
+    } else {
+        dev->regs.clock.clk_source &= ~CLCSOURCE_CRYSTALEN;
+    }
+    if (dev->cfg.no_pll) {
+        dev->regs.clock.clk_source |= CLCSOURCE_PLLBYPASS;
+    } else {
+        dev->regs.clock.clk_source &= ~CLCSOURCE_PLLBYPASS;
+    }
+    if (dev->cfg.allow_mul_4) {
+        dev->regs.clock.brg_config = BRGCONFIG_4XMODE;
+    } else if (dev->cfg.allow_mul_2) {
+        dev->regs.clock.brg_config = BRGCONFIG_2XMODE;
+    } else {
+        dev->regs.clock.brg_config = 0;
+    }
+    max3107_compute_clock_config(dev->cfg.osc_freq, conf->uc_speed, &dev->regs.clock);
```

Comment:
You are returning `best_br` here, should we be using it instead for our config or something ? Do we need to verify it with our specified config.

Reply:
Other drivers silently set register values from requested baud rate. NRF5x for example outlines in product specification what is the actual baud rate for requested one. For MCUs, code calculates it in much simpler way since there is only one clock involved (or takes values from table) and you don't get information what is the actual speed at all.
In MAX3107 case which external clock/oscillator, code finds most accurate settings and returns it so user can verify it during the development phase by just reading return value of this functions instead of checking registers and computing it manually.
So this value is informative only. For example on my dev board that has 36868400 clock I could get as high as 19,7 MHz and not 24MHz that seems to be possible from MAX3107 datasheet.

Reply:
I see, makes sense. some comment or a log statement will be helpful in understanding what the real Vs specified config is.

Reply:
Separate function to get actual baudrate added